auditbeat/module/file_integrity/kprobes/path.go

```diff
@@ -0,0 +1,324 @@
+		return nil, err
+	}
+	names, err := f.Readdirnames(-1)
+	_ = f.Close()
+	if err != nil {
+		return nil, err
+	}
+	sort.Strings(names)
+	return names, nil
+}
+
+func (r *pTraverser) ErrC() <-chan error {
+	return r.errC
+}
+
+func (r *pTraverser) WalkAsync(path string, depth uint32, tid uint32) {
+	if r.ctx.Err() != nil {
+		return
+	}
+
+	go func() {
+		walkErr := r.e.Run(func() error {
+			return r.walk(r.ctx, path, depth, true, tid)
+		})
+
+		if walkErr == nil {
+			return
+		}
+
+		select {
+		case r.errC <- walkErr:
+		case <-r.ctx.Done():
+		}
+
+	}()
+}
+
+func (r *pTraverser) walkRecursive(ctx context.Context, path string, mounts mountPoints, depth uint32, isFromMove bool, tid uint32) error {
+
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	if r.ctx.Err() != nil {
+		return r.ctx.Err()
+	}
+
+	if !r.isRecursive && depth > 1 {
+		return nil
+	}
+
+	// get the mountpoint associated to this path
+	mnt := mounts.getMountByPath(path)
+	if mnt == nil {
+		return fmt.Errorf("could not find mount for %s", path)
+	}
+
+	// add the inotify watcher if it does not exist
+	if _, err := r.w.Add(mnt.DeviceMajor, mnt.DeviceMinor, path); err != nil {
+		return err
+	}
+
+	r.mtx.Lock()
+	info, err := lstat(path)
+	if err != nil {
+		// maybe this path got deleted/moved in the meantime
+		// return nil
+		r.mtx.Unlock()
+		//lint:ignore nilerr no errors returned for lstat from walkRecursive
+		return nil
+	}
+
+	// if we are about to stat the root of the mountpoint, and the subtree has a different base
+	// from the base of the path (e.g. /watch [path] -> /etc/test [subtree])
+	// the filename reported in the kprobe event will be "test" instead of "watch". Thus, we need to
+	// construct the filename based on the base name of the subtree.
+	mntPath := strings.Replace(path, mnt.Path, "", 1)
+	if !strings.HasPrefix(mntPath, mnt.Subtree) {
+		mntPath = filepath.Join(mnt.Subtree, mntPath)
+	}
+
+	matchFileName := filepath.Base(mntPath)
+
+	r.statQueue = append(r.statQueue, statMatch{
+		ino:        info.Sys().(*syscall.Stat_t).Ino,
+		major:      mnt.DeviceMajor,
+		minor:      mnt.DeviceMinor,
+		depth:      depth,
+		fileName:   matchFileName,
+		isFromMove: isFromMove,
+		tid:        tid,
+		fullPath:   path,
+	})
+	r.mtx.Unlock()
+
+	if !info.IsDir() {
+		return nil
+	}
+
+	names, err := readDirNames(path)
+	if err != nil {
+		// maybe this dir got deleted/moved in the meantime
+		// return nil
+		//lint:ignore nilerr no errors returned for readDirNames from walkRecursive
+		return nil
+	}
+
+	for _, name := range names {
+		filename := filepath.Join(path, name)
+		if err = r.walkRecursive(ctx, filename, mounts, depth+1, isFromMove, tid); err != nil {
+			//lint:ignore nilerr no errors returned for readDirNames from walkRecursive
+			return nil
+		}
+	}
+	return nil
+}
+
+func (r *pTraverser) waitForWalk(ctx context.Context) error {
+	r.mtx.Lock()
+
+	// statQueue is already empty, return
+	if len(r.statQueue) == 0 {
+		r.mtx.Unlock()
```

Comment:
can this Unlock() be deferred?

Reply:
I mean sure this particular one can be deferred but why we return right after it

Reply:
similar to the above you can't hold the lock until the function exits since the same mutex is used by `GetMonitorPath` which is used to "empty" the statQueue